csrc/device_lower/pass/circular_buffer.cpp

```diff
@@ -753,46 +743,50 @@ class CloneTmaCircularBufferLoopAndInsertSync
     kir::TensorIndex* stage_mbarrier =
         IrBuilder::create<kir::TensorIndex>(all_mbarriers, loop_index);
 
-    // Get mbarrier_token for this circular buffer stage.
-    TensorView* all_mbarrier_tokens =
-        GpuLower::current()->tmaCircularBufferInfo().getMBarrierToken(ldst);
-    kir::TensorIndex* stage_token =
-        IrBuilder::create<kir::TensorIndex>(all_mbarrier_tokens, loop_index);
-
     Val* tx_count = GpuLower::current()->commonScalarMap().hoistScalar(
         getSizeOfTmaLoad(ldst), for_loop_stack_);
     kir::MBarrierArriveExpectTx* mbarrier_arrive_tx =
         IrBuilder::create<kir::MBarrierArriveExpectTx>(
-            stage_token, stage_mbarrier, tx_count);
+            nullptr, stage_mbarrier, tx_count);
 
     return mbarrier_arrive_tx;
   }
 
-  // This function creates kir::MBarrierWait for given LoadStoreOp and circular
-  // buffer stage.
-  kir::MBarrierWait* createMbarrierWait(LoadStoreOp* ldst, Val* loop_index) {
+  // This function creates kir::MBarrierWaitParity for given LoadStoreOp and
+  // circular buffer stage.
+  kir::MBarrierWaitParity* createMbarrierWait(
+      LoadStoreOp* ldst,
+      Val* stage,
+      Val* loop_index) {
     NVF_ERROR(ldst != nullptr);
     NVF_ERROR(loop_index != nullptr);
 
-    // Get mbarrier_token for this circular buffer stage.
+    int64_t stage_depth =
+        GpuLower::current()->circularBufferInfo().getStageDepthFor(
+            circular_buffer_loop_->iter_domain());
+
+    // Get mbarrier for this circular buffer stage.
     TensorView* all_mbarriers = GpuLower::current()->ldstMBarrierMap().at(ldst);
     kir::TensorIndex* stage_mbarrier =
-        IrBuilder::create<kir::TensorIndex>(all_mbarriers, loop_index);
-
-    // Get mbarrier_token for this circular buffer stage.
-    TensorView* all_mbarrier_tokens =
-        GpuLower::current()->tmaCircularBufferInfo().getMBarrierToken(ldst);
-    kir::TensorIndex* stage_token =
-        IrBuilder::create<kir::TensorIndex>(all_mbarrier_tokens, loop_index);
-
-    kir::MBarrierWait* mbarrier_wait =
-        IrBuilder::create<kir::MBarrierWait>(stage_mbarrier, stage_token);
+        IrBuilder::create<kir::TensorIndex>(all_mbarriers, stage);
+
+    // Get mbarrier_parity for this circular buffer stage.
+    auto depth = IrBuilder::create<Val>(stage_depth, DataType::UInt32);
+    auto two = IrBuilder::create<Val>(2, DataType::UInt32);
```

Comment:
Comment about stage parity:

We have an mbarrier for each circular buffer stage, so `loop_index / stage_depth` is `loop_index_per_stage`.
The valid values of phaseParity operand are 0 and 1, so we take the modulo of `loop_index_per_stage` with a divisor of 2. See https://docs.nvidia.com/cuda/parallel-thread-execution/#parallel-synchronization-and-communication-instructions-mbarrier-test-wait-mbarrier-try-wait for reference.